File: Problem-Set/Graph_Algorithms/COMM3/q45_tim.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include<math.h>
int main(){
    int t;
    scanf("%d",&t);
    while (t>0) {
        int n,x1,y1,x2,y2,x3,y3;
        double a,b,c;
        scanf("%d %d %d %d %d %d %d",&n,&x1,&y1,&x2,&y2,&x3,&y3);
        a=sqrt((x2-x1)*(x2-x1) + (y2-y1)*(y2-y1));
        b=sqrt((x3-x1)*(x3-x1) + (y3-y1)*(y3-y1));
        c=sqrt((x2-x3)*(x2-x3) + (y2-y3)*(y2-y3));
        if(c<=b && a<=b){
            if(c<=n && a<=n)
                printf("yes\n");
            else printf("no\n");
         }
        else if(a<=c && b<=c){
            if(a<=n && b<=n)
                printf("yes\n");
            else printf("no\n");
        }
        else if(c<=a && b<=a){
            if(b<=n && c<=n)
                printf("yes\n");
            else printf("no\n");
        }
        else printf("no\n");
        t--;
    }
}

File: Problem-Set/Graph_Algorithms/COMM3/q45_tim.hs
import System.IO
import Control.Monad
main = readLn >>= flip replicateM_ runOneCase
runOneCase :: IO ()
runOneCase = do
	r <- readLn
	a <- liftM (map read . words) $ getLine
	b <- liftM (map read . words) $ getLine
	c <- liftM (map read . words) $ getLine
	putStrLn (if checkDists a b c r then "yes" else "no")
checkDists :: [Float] -> [Float] -> [Float] -> Integer -> Bool
checkDists a b c r = atLeastTwo (dist a b) (dist a c) (dist b c) r
dist :: [Float] -> [Float] -> Float
dist a b = sqrt ((a !! 0 - b !! 0) ^ 2 + (a !! 1 - b !! 1) ^ 2)
atLeastTwo :: Float -> Float -> Float -> Integer -> Bool
atLeastTwo a b c r = count True [a <= fr, b <= fr, c <= fr] >= 2
	where
		fr = fromInteger r + 0.01
		count v = foldl (\acc x -> if v == x then acc + 1 else acc) 0
